File: runner-go/runner.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"os"
	"runtime"
	"sort"
	"time"
)

type CaseResult struct {
	Case            string         `json:"case"`
	Param           ParamEntry     `json:"param"`
	ItersMs         []float64      `json:"iters_ms"`
	MedianMs        float64        `json:"median_ms"`
	MinMs           float64        `json:"min_ms"`
	MaxMs           float64        `json:"max_ms"`
	MeanMs          float64        `json:"mean_ms"`
	ThroughputMBps  *float64       `json:"throughput_mbps,omitempty"`
	PeakRSSMB       float64        `json:"peak_rss_mb"`
	Notes           string         `json:"notes,omitempty"`
	Skipped         bool           `json:"skipped,omitempty"`
	SkipReason      string         `json:"skip_reason,omitempty"`
}

type RunResult struct {
	Runner          string       `json:"runner"`
	ClientVersion   string       `json:"client_version"`
	BeeVersion      string       `json:"bee_version"`
	BenchSpecHash   string       `json:"bench_spec_hash"`
	StartedAt       string       `json:"started_at"`
	Host            HostInfo     `json:"host"`
	BeeURL          string       `json:"bee_url"`
	BatchID         string       `json:"batch_id"`
	Iters           int          `json:"iters"`
	Results         []CaseResult `json:"results"`
}

type HostInfo struct {
	OS    string `json:"os"`
	Arch  string `json:"arch"`
	CPU   string `json:"cpu"`
	NumCPU int   `json:"num_cpu"`
}

// caseFn is the signature of a benchmark case implementation.
//   - returns durations in milliseconds for each iteration
//   - throughput-bytes is the total bytes processed per iter (for MB/s calc); 0 if N/A
//   - notes is a free-form string
type caseFn func(ctx context.Context, p ParamEntry) (ms []float64, bytesPerIter int64, notes string, err error)

// runCase orchestrates warmup + iters + RSS sampling around a caseFn.
func runCase(ctx context.Context, env *Env, c CaseSpec, p ParamEntry, fn caseFn) CaseResult {
	res := CaseResult{Case: c.ID, Param: p}

	// Warmup: invoke fn with reduced iteration count if possible. We expose a
	// "warmup" boolean in the param map so the case impl can skip RSS work etc.
	warmup := env.Spec.WarmupNet
	if c.Kind == "cpu" {
		warmup = env.Spec.WarmupCPU
	}
	if warmup > 0 {
		warmupParam := copyParam(p)
		warmupParam["warmup"] = true
		warmupParam["count_override"] = warmup // case impls may use this to scale down
		_, _, _, _ = fn(ctx, warmupParam)
	}

	// Timed run
	sampler := startRSSSampler(env.Spec.RSSSampleIntervalMs)
	ms, bytesPerIter, notes, err := fn(ctx, p)
	res.PeakRSSMB = roundTo(sampler.stopAndPeakMB(), 1)

	if err != nil {
		res.Notes = "ERR: " + err.Error()
		return res
	}
	if len(ms) == 0 {
		res.Skipped = true
		// If the impl signaled a SKIP via notes, surface that reason.
		if len(notes) > 6 && notes[:6] == "SKIP: " {
			res.SkipReason = notes[6:]
		} else if notes != "" {
			res.SkipReason = notes
		} else {
			res.SkipReason = "no iterations recorded"
		}
		return res
	}
	res.ItersMs = roundSlice(ms, 3)
	res.MinMs, res.MaxMs, res.MedianMs, res.MeanMs = stats(ms)
	if bytesPerIter > 0 {
		// throughput = bytes / median_seconds / 1MB
		mb := float64(bytesPerIter) / (1024 * 1024)
		t := mb / (res.MedianMs / 1000.0)
		t = roundTo(t, 2)
		res.ThroughputMBps = &t
	}
	res.Notes = notes
	return res
}

func copyParam(p ParamEntry) ParamEntry {
	out := make(ParamEntry, len(p))
	for k, v := range p {
		out[k] = v
	}
	return out
}

func stats(ms []float64) (min, max, median, mean float64) {
	if len(ms) == 0 {
		return
	}
	sorted := append([]float64(nil), ms...)
	sort.Float64s(sorted)
	min, max = sorted[0], sorted[len(sorted)-1]
	if len(sorted)%2 == 1 {
		median = sorted[len(sorted)/2]
	} else {
		median = (sorted[len(sorted)/2-1] + sorted[len(sorted)/2]) / 2
	}
	var sum float64
	for _, v := range ms {
		sum += v
	}
	mean = sum / float64(len(ms))
	return roundTo(min, 3), roundTo(max, 3), roundTo(median, 3), roundTo(mean, 3)
}

func roundTo(v float64, places int) float64 {
	if math.IsNaN(v) || math.IsInf(v, 0) {
		return v
	}
	mult := math.Pow(10, float64(places))
	return math.Round(v*mult) / mult
}

func roundSlice(in []float64, places int) []float64 {
	out := make([]float64, len(in))
	for i, v := range in {
		out[i] = roundTo(v, places)
	}
	return out
}

// time-this helper for case implementations.
func timeIt(fn func() error) (float64, error) {
	start := time.Now()
	err := fn()
	return float64(time.Since(start).Microseconds()) / 1000.0, err
}

func cooldown(sec int) {
	if sec > 0 {
		time.Sleep(time.Duration(sec) * time.Second)
	}
}

func writeResult(path string, r *RunResult) error {
	if err := os.MkdirAll(dirOf(path), 0o755); err != nil {
		return err
	}
	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()
	enc := json.NewEncoder(f)
	enc.SetIndent("", "  ")
	return enc.Encode(r)
}

func dirOf(p string) string {
	for i := len(p) - 1; i >= 0; i-- {
		if p[i] == '/' {
			return p[:i]
		}
	}
	return "."
}

func hostInfo() HostInfo {
	return HostInfo{
		OS:     runtime.GOOS,
		Arch:   runtime.GOARCH,
		CPU:    cpuModel(),
		NumCPU: runtime.NumCPU(),
	}
}

func cpuModel() string {
	raw, err := os.ReadFile("/proc/cpuinfo")
	if err != nil {
		return ""
	}
	for _, line := range splitLines(string(raw)) {
		if len(line) < 11 || line[:10] != "model name" {
			continue
		}
		// "model name\t: Intel(R) ..."
		for i := 0; i < len(line); i++ {
			if line[i] == ':' && i+2 < len(line) {
				return line[i+2:]
			}
		}
	}
	return ""
}

func splitLines(s string) []string {
	var out []string
	start := 0
	for i := 0; i < len(s); i++ {
		if s[i] == '\n' {
			out = append(out, s[start:i])
			start = i + 1
		}
	}
	if start < len(s) {
		out = append(out, s[start:])
	}
	return out
}

// debug helper — print to stderr without buffering
func logf(f string, args ...any) {
	fmt.Fprintf(os.Stderr, f+"\n", args...)
}

File: runner-rs/src/runner.rs
use serde::Serialize;
use std::time::Instant;

use crate::spec::{BenchSpec, CaseSpec, ParamEntry};

pub const RUNNER_NAME: &str = "rs";

#[derive(Serialize, Default)]
pub struct CaseResult {
    #[serde(rename = "case")]
    pub case_id: String,
    pub param: ParamEntry,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub iters_ms: Vec<f64>,
    pub median_ms: f64,
    pub min_ms: f64,
    pub max_ms: f64,
    pub mean_ms: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub throughput_mbps: Option<f64>,
    pub peak_rss_mb: f64,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub notes: String,
    #[serde(skip_serializing_if = "is_false")]
    pub skipped: bool,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub skip_reason: String,
}

fn is_false(b: &bool) -> bool { !*b }

#[derive(Serialize)]
pub struct RunResult {
    pub runner: String,
    pub client_version: String,
    pub bee_version: String,
    pub bench_spec_hash: String,
    pub started_at: String,
    pub host: HostInfo,
    pub bee_url: String,
    pub batch_id: String,
    pub iters: usize,
    pub results: Vec<CaseResult>,
}

#[derive(Serialize, Default)]
pub struct HostInfo {
    pub os: String,
    pub arch: String,
    pub cpu: String,
    pub num_cpu: usize,
}

pub fn host_info() -> HostInfo {
    HostInfo {
        os: std::env::consts::OS.to_string(),
        arch: std::env::consts::ARCH.to_string(),
        cpu: cpu_model(),
        num_cpu: std::thread::available_parallelism().map(|n| n.get()).unwrap_or(0),
    }
}

fn cpu_model() -> String {
    let raw = match std::fs::read_to_string("/proc/cpuinfo") {
        Ok(s) => s,
        Err(_) => return String::new(),
    };
    for line in raw.lines() {
        if let Some(idx) = line.find(':') {
            if line.starts_with("model name") {
                return line[idx + 1..].trim().to_string();
            }
        }
    }
    String::new()
}

pub fn stats(ms: &[f64]) -> (f64, f64, f64, f64) {
    if ms.is_empty() { return (0.0, 0.0, 0.0, 0.0); }
    let mut sorted = ms.to_vec();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let min = sorted[0];
    let max = *sorted.last().unwrap();
    let median = if sorted.len() % 2 == 1 {
        sorted[sorted.len() / 2]
    } else {
        (sorted[sorted.len() / 2 - 1] + sorted[sorted.len() / 2]) / 2.0
    };
    let mean = ms.iter().sum::<f64>() / ms.len() as f64;
    (round3(min), round3(max), round3(median), round3(mean))
}

pub fn round3(v: f64) -> f64 { (v * 1000.0).round() / 1000.0 }
pub fn round2(v: f64) -> f64 { (v * 100.0).round() / 100.0 }
pub fn round1(v: f64) -> f64 { (v * 10.0).round() / 10.0 }

pub fn time_it<F, T>(f: F) -> (f64, T) where F: FnOnce() -> T {
    let start = Instant::now();
    let r = f();
    let ms = start.elapsed().as_micros() as f64 / 1000.0;
    (ms, r)
}

pub async fn time_async<F, Fut, T>(f: F) -> (f64, T)
where F: FnOnce() -> Fut, Fut: std::future::Future<Output = T>,
{
    let start = Instant::now();
    let r = f().await;
    let ms = start.elapsed().as_micros() as f64 / 1000.0;
    (ms, r)
}

pub fn finalize_result(
    case: &CaseSpec,
    p: ParamEntry,
    ms: Vec<f64>,
    bytes_per_iter: i64,
    notes: String,
    peak_rss_mb: f64,
) -> CaseResult {
    let mut r = CaseResult {
        case_id: case.id.clone(),
        param: p,
        peak_rss_mb: round1(peak_rss_mb),
        ..Default::default()
    };
    if ms.is_empty() {
        r.skipped = true;
        if let Some(rest) = notes.strip_prefix("SKIP: ") {
            r.skip_reason = rest.to_string();
        } else if !notes.is_empty() {
            r.skip_reason = notes;
        } else {
            r.skip_reason = "no iterations recorded".to_string();
        }
        return r;
    }
    let (min, max, median, mean) = stats(&ms);
    r.iters_ms = ms.iter().map(|v| round3(*v)).collect();
    r.min_ms = min;
    r.max_ms = max;
    r.median_ms = median;
    r.mean_ms = mean;
    if bytes_per_iter > 0 {
        let mb = bytes_per_iter as f64 / (1024.0 * 1024.0);
        let t = mb / (median / 1000.0);
        r.throughput_mbps = Some(round2(t));
    }
    r.notes = notes;
    r
}

pub fn iters_for(_spec: &BenchSpec, _p: &ParamEntry) -> usize { 5 }
